File: mobile/widgets/ios/MomentWidgets.swift
import WidgetKit
import SwiftUI

// Candle-style glance widgets. They share the cached GET /widget/summary
// payload that WidgetDataLoader already fetches, decoded here into the extra
// fields (partner mood, days together, anniversary, next date, love note,
// secret flag, today's question). Each opens the app at the matching screen
// through a loversrock:// deep link.

private extension Color {
    static let mAccent = Color(red: 0.91, green: 0.38, blue: 0.48)
    static let mText = Color(red: 0.17, green: 0.14, blue: 0.13)
    static let mMuted = Color(red: 0.55, green: 0.50, blue: 0.47)
}

struct MomentSummary: Decodable {
    struct Mood: Decodable { let emoji: String; let text: String? }
    struct Anniversary: Decodable { let date: String; let daysRemaining: Int; let years: Int }
    struct NextDate: Decodable { let title: String; let daysRemaining: Int }
    struct Note: Decodable { let title: String?; let body: String }
    struct Countdown: Decodable { let label: String; let daysRemaining: Int }

    let paired: Bool
    let streakCount: Int
    let promptAnsweredToday: Bool
    let streakFreezes: Int?
    let partnerName: String?
    let partnerMood: Mood?
    let daysTogether: Int?
    let anniversary: Anniversary?
    let nextDate: NextDate?
    let latestNote: Note?
    let secretMessageWaiting: Bool?
    let todayQuestion: String?
    let nextCountdown: Countdown?
    let distanceKm: Double?

    static let preview = MomentSummary(
        paired: true, streakCount: 42, promptAnsweredToday: false, streakFreezes: 1, partnerName: "Amara",
        partnerMood: Mood(emoji: "🥰", text: "Thinking of you"), daysTogether: 1234,
        anniversary: Anniversary(date: "2026-02-14", daysRemaining: 142, years: 4),
        nextDate: NextDate(title: "Sunset picnic", daysRemaining: 3),
        latestNote: Note(title: nil, body: "Don't forget: dinner at 8, wear the blue shirt ❤️"),
        secretMessageWaiting: true, todayQuestion: "What's a small thing I do that makes your day?",
        nextCountdown: Countdown(label: "Trip to Lisbon", daysRemaining: 12), distanceKm: 8.4
    )
}

enum MomentLoader {
    static var cached: MomentSummary? {
        guard let data = UserDefaults(suiteName: WidgetDataLoader.appGroup)?.data(forKey: "cachedSummary") else { return nil }
        return try? JSONDecoder().decode(MomentSummary.self, from: data)
    }

    static func load() async -> MomentSummary? {
        _ = await WidgetDataLoader.fetch() // refreshes the shared cache
        return cached
    }
}

enum MomentKind: String, CaseIterable {
    case daysTogether, anniversary, streak, countdown, distance, partnerMood, nextDate, loveNote, secretMessage, dailyQuestion, quickKiss

    var displayName: String {
        switch self {
        case .daysTogether: return "Days together"
        case .anniversary: return "Anniversary"
        case .streak: return "Streak"
        case .countdown: return "Countdown"
        case .distance: return "Distance"
        case .partnerMood: return "Partner mood"
        case .nextDate: return "Next date"
        case .loveNote: return "Love note"
        case .secretMessage: return "Secret message"
        case .dailyQuestion: return "Daily question"
        case .quickKiss: return "Quick kiss"
        }
    }

    var deepLink: URL {
        let path: String
        switch self {
        case .daysTogether, .anniversary: path = "home"
        case .streak: path = "streak"
        case .countdown: path = "countdowns"
        case .distance: path = "distance"
        case .partnerMood: path = "mood"
        case .nextDate: path = "dates"
        case .loveNote: path = "notes"
        case .secretMessage: path = "secret"
        case .dailyQuestion: path = "question"
        case .quickKiss: path = "thumbkiss"
        }
        return URL(string: "loversrock://\(path)")!
    }

    /// Lock screen families for the ones that read well in a tiny space.
    var families: [WidgetFamily] {
        switch self {
        case .daysTogether, .anniversary, .streak, .partnerMood, .secretMessage, .countdown:
            return [.systemSmall, .accessoryCircular, .accessoryRectangular, .accessoryInline]
        case .loveNote, .dailyQuestion:
            return [.systemSmall, .systemMedium, .accessoryRectangular]
        case .quickKiss:
            return [.systemSmall, .accessoryCircular]
        default:
            return [.systemSmall, .accessoryRectangular]
        }
    }
}

struct MomentContent {
    let icon: String
    let value: String
    let label: String
    var sub: String = ""
    var long: Bool = false
}

func momentContent(_ kind: MomentKind, _ s: MomentSummary?, signedIn: Bool) -> MomentContent {
    if kind == .quickKiss { return MomentContent(icon: "💋", value: "Thumb Kiss", label: "Tap to connect") }
    guard signedIn else { return MomentContent(icon: "🕯️", value: "loversrock.", label: "Sign in to set up") }
    guard let s else { return MomentContent(icon: "🕯️", value: "loversrock.", label: "Open the app") }
    guard s.paired else { return MomentContent(icon: "💞", value: "Not paired", label: "Pair in the app") }
    let partner = s.partnerName ?? "Your partner"

    switch kind {
    case .daysTogether:
        return MomentContent(icon: "❤️", value: "\(s.daysTogether ?? 0)", label: "days together", sub: "with \(partner)")
    case .anniversary:
        guard let a = s.anniversary else { return MomentContent(icon: "💍", value: "—", label: "Set your anniversary") }
        return MomentContent(icon: "💍", value: a.daysRemaining == 0 ? "Today!" : "\(a.daysRemaining)d", label: "until our anniversary", sub: a.years > 0 ? "\(a.years) years" : "")
    case .streak:
        let sub = !s.promptAnsweredToday ? "Question waiting" : ((s.streakFreezes ?? 0) > 0 ? "❄️ \(s.streakFreezes ?? 0) banked" : "All caught up")
        return MomentContent(icon: "🔥", value: "\(s.streakCount)", label: "day streak", sub: sub)
    case .countdown:
        guard let c = s.nextCountdown else { return MomentContent(icon: "⏳", value: "—", label: "No countdown yet") }
        return MomentContent(icon: "⏳", value: "\(c.daysRemaining)d", label: c.label)
    case .distance:
        guard let km = s.distanceKm else { return MomentContent(icon: "📍", value: "—", label: "Sharing off") }
        return MomentContent(icon: "📍", value: km < 1 ? "<1 km" : "\(Int(km)) km", label: "apart", sub: "from \(partner)")
    case .partnerMood:
        guard let m = s.partnerMood else { return MomentContent(icon: "💭", value: partner, label: "hasn't set a mood") }
        return MomentContent(icon: m.emoji, value: partner, label: "is feeling", sub: m.text ?? "")
    case .nextDate:
        guard let d = s.nextDate else { return MomentContent(icon: "📅", value: "No date planned", label: "Swipe for ideas", long: true) }
        return MomentContent(icon: "📅", value: d.title, label: d.daysRemaining == 0 ? "today" : "in \(d.daysRemaining) days", long: true)
    case .loveNote:
        guard let n = s.latestNote else { return MomentContent(icon: "💌", value: "No notes yet", label: "from \(partner)", long: true) }
        return MomentContent(icon: "💌", value: n.body, label: "from \(partner)", sub: n.title ?? "", long: true)
    case .secretMessage:
        // Only ever that a secret exists — never its words.
        return (s.secretMessageWaiting ?? false)
            ? MomentContent(icon: "💌", value: "You have a message ❤️", label: "Tap to open", long: true)
            : MomentContent(icon: "🔒", value: "No new secrets", label: "Leave one for \(partner)", long: true)
    case .dailyQuestion:
        return MomentContent(icon: "💬", value: s.todayQuestion ?? "No question today", label: "Today's question", sub: s.promptAnsweredToday ? "Answered ✓" : "Tap to answer", long: true)
    case .quickKiss:
        return MomentContent(icon: "💋", value: "Thumb Kiss", label: "Tap to connect")
    }
}

struct MomentEntry: TimelineEntry {
    let date: Date
    let summary: MomentSummary?
    let signedIn: Bool
}

struct MomentProvider: TimelineProvider {
    func placeholder(in context: Context) -> MomentEntry { MomentEntry(date: Date(), summary: .preview, signedIn: true) }

    func getSnapshot(in context: Context, completion: @escaping (MomentEntry) -> Void) {
        if context.isPreview { completion(placeholder(in: context)); return }
        completion(MomentEntry(date: Date(), summary: MomentLoader.cached, signedIn: WidgetDataLoader.credentials != nil))
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<MomentEntry>) -> Void) {
        Task {
            let summary = await MomentLoader.load()
            let entry = MomentEntry(date: Date(), summary: summary, signedIn: WidgetDataLoader.credentials != nil)
            let next = Calendar.current.date(byAdding: .minute, value: 30, to: Date()) ?? Date().addingTimeInterval(1800)
            completion(Timeline(entries: [entry], policy: .after(next)))
        }
    }
}

struct MomentView: View {
    @Environment(\.widgetFamily) var family
    let kind: MomentKind
    let entry: MomentEntry

    var body: some View {
        let c = momentContent(kind, entry.summary, signedIn: entry.signedIn)
        Group {
            switch family {
            case .accessoryInline:
                Text("\(c.icon) \(c.value) \(c.long ? "" : c.label)")
            case .accessoryCircular:
                ZStack {
                    AccessoryWidgetBackground()
                    VStack(spacing: 0) {
                        Text(c.icon).font(.caption2)
                        Text(c.long ? "•" : c.value).font(.system(size: 14, weight: .bold)).minimumScaleFactor(0.5).lineLimit(1)
                    }
                }
            case .accessoryRectangular:
                VStack(alignment: .leading, spacing: 1) {
                    Text("\(c.icon) \(c.long ? c.label : c.value)").font(.headline).lineLimit(1)
                    Text(c.long ? c.value : c.label).font(.caption2).lineLimit(2)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
            default:
                VStack(spacing: 4) {
                    Text(c.icon).font(.system(size: 28))
                    Text(c.value)
                        .font(c.long ? .system(size: 14, weight: .semibold) : .system(size: 28, weight: .bold))
                        .foregroundColor(c.long ? .mText : .mAccent)
                        .multilineTextAlignment(.center)
                        .lineLimit(c.long ? 4 : 1)
                        .minimumScaleFactor(0.6)
                    Text(c.label).font(.caption).foregroundColor(.mText).lineLimit(2).multilineTextAlignment(.center)
                    if !c.sub.isEmpty { Text(c.sub).font(.caption2).foregroundColor(.mMuted).lineLimit(1) }
                }
                .padding(10)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            }
        }
        .widgetURL(kind.deepLink)
    }
}

private func momentConfiguration(_ kind: MomentKind) -> some WidgetConfiguration {
    StaticConfiguration(kind: "LoversRockMoment.\(kind.rawValue)", provider: MomentProvider()) { entry in
        if #available(iOS 17.0, *) {
            MomentView(kind: kind, entry: entry).containerBackground(.background, for: .widget)
        } else {
            MomentView(kind: kind, entry: entry)
        }
    }
    .configurationDisplayName(kind.displayName)
    .description("loversrock — \(kind.displayName.lowercased())")
    .supportedFamilies(kind.families)
}

struct DaysTogetherWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.daysTogether) } }
struct AnniversaryWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.anniversary) } }
struct StreakWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.streak) } }
struct CountdownWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.countdown) } }
struct DistanceWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.distance) } }
struct PartnerMoodWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.partnerMood) } }
struct NextDateWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.nextDate) } }
struct LoveNoteWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.loveNote) } }
struct SecretMessageWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.secretMessage) } }
struct DailyQuestionWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.dailyQuestion) } }
struct QuickKissWidget: Widget { var body: some WidgetConfiguration { momentConfiguration(.quickKiss) } }

// MARK: - Canvas widget: the shared drawing, rendered from vector strokes.

struct CanvasStroke: Decodable {
    struct Point: Decodable { let x: Double; let y: Double }
    let color: String?
    let width: Double?
    let tool: String?
    let points: [Point]
}

struct CanvasPayload: Decodable {
    let strokes: [CanvasStroke]
    let background: String
}

private func hexColor(_ hex: String?) -> Color {
    guard let hex, hex.hasPrefix("#"), hex.count >= 7, let v = Int(hex.dropFirst().prefix(6), radix: 16) else { return .mAccent }
    return Color(red: Double((v >> 16) & 255) / 255, green: Double((v >> 8) & 255) / 255, blue: Double(v & 255) / 255)
}

struct CanvasEntry: TimelineEntry {
    let date: Date
    let canvas: CanvasPayload?
}

struct CanvasProvider: TimelineProvider {
    func placeholder(in context: Context) -> CanvasEntry { CanvasEntry(date: Date(), canvas: nil) }
    func getSnapshot(in context: Context, completion: @escaping (CanvasEntry) -> Void) { completion(placeholder(in: context)) }

    func getTimeline(in context: Context, completion: @escaping (Timeline<CanvasEntry>) -> Void) {
        Task {
            var payload: CanvasPayload?
            if let creds = WidgetDataLoader.credentials, let url = URL(string: "\(creds.apiUrl)/widget/canvas") {
                var request = URLRequest(url: url)
                request.setValue(creds.token, forHTTPHeaderField: "X-Widget-Token")
                request.timeoutInterval = 10
                if let (data, response) = try? await URLSession.shared.data(for: request),
                   (response as? HTTPURLResponse)?.statusCode == 200 {
                    payload = try? JSONDecoder().decode(CanvasPayload.self, from: data)
                }
            }
            let next = Calendar.current.date(byAdding: .minute, value: 30, to: Date()) ?? Date().addingTimeInterval(1800)
            completion(Timeline(entries: [CanvasEntry(date: Date(), canvas: payload)], policy: .after(next)))
        }
    }
}

struct CanvasWidgetView: View {
    let entry: CanvasEntry
    var body: some View {
        Group {
            if let canvas = entry.canvas, !canvas.strokes.isEmpty {
                GeometryReader { geo in
                    // Keep the 3:4 page, centred.
                    let h = min(geo.size.height, geo.size.width * 4 / 3)
                    let w = h * 3 / 4
                    Canvas { ctx, _ in
                        ctx.fill(Path(CGRect(x: 0, y: 0, width: w, height: h)), with: .color(hexColor(canvas.background)))
                        for s in canvas.strokes {
                            var path = Path()
                            for (i, p) in s.points.enumerated() {
                                let pt = CGPoint(x: p.x * w, y: p.y * h)
                                if i == 0 { path.move(to: pt) } else { path.addLine(to: pt) }
                            }
                            let color = s.tool == "eraser" ? hexColor(canvas.background) : hexColor(s.color)
                            ctx.stroke(path, with: .color(color), style: StrokeStyle(lineWidth: max(1, (s.width ?? 0.008) * w), lineCap: .round, lineJoin: .round))
                        }
                    }
                    .frame(width: w, height: h)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                }
            } else {
                Text("Draw something together 🎨").font(.caption).foregroundColor(.mMuted).multilineTextAlignment(.center).padding()
            }
        }
        .widgetURL(URL(string: "loversrock://canvas"))
    }
}

struct CanvasWidget: Widget {
    var body: some WidgetConfiguration {
        StaticConfiguration(kind: "LoversRockCanvas", provider: CanvasProvider()) { entry in
            if #available(iOS 17.0, *) {
                CanvasWidgetView(entry: entry).containerBackground(.background, for: .widget)
            } else {
                CanvasWidgetView(entry: entry)
            }
        }
        .configurationDisplayName("Canvas")
        .description("Your shared drawing.")
        .supportedFamilies([.systemSmall, .systemLarge])
    }
}

/// Grouped so the main bundle stays under WidgetKit's per-builder limit.
struct MomentWidgetsA: WidgetBundle {
    var body: some Widget {
        DaysTogetherWidget()
        AnniversaryWidget()
        StreakWidget()
        CountdownWidget()
        DistanceWidget()
        PartnerMoodWidget()
    }
}

struct MomentWidgetsB: WidgetBundle {
    var body: some Widget {
        NextDateWidget()
        LoveNoteWidget()
        SecretMessageWidget()
        DailyQuestionWidget()
        QuickKissWidget()
        CanvasWidget()
    }
}

File: mobile/widgets/android/MomentWidgets.kt
package com.loversrock.app.widgets

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.TypedValue
import android.view.View
import android.widget.RemoteViews
import com.loversrock.app.R
import org.json.JSONObject

/**
 * The Candle-style glance widgets. Each is a tiny subclass of
 * MomentWidgetProvider naming its Kind; they share one layout
 * (widget_moment.xml), one data source (MomentRepository) and open the app
 * at the matching screen via a loversrock:// deep link.
 */
enum class MomentKind(val deepLink: String) {
    DAYS_TOGETHER("home"),
    ANNIVERSARY("home"),
    STREAK("streak"),
    COUNTDOWN("countdowns"),
    DISTANCE("distance"),
    PARTNER_MOOD("mood"),
    NEXT_DATE("dates"),
    LOVE_NOTE("notes"),
    SECRET_MESSAGE("secret"),
    DAILY_QUESTION("question"),
    QUICK_KISS("thumbkiss"),
}

data class MomentContent(val icon: String, val value: String, val label: String, val sub: String = "", val longValue: Boolean = false)

object MomentFormatter {
    fun content(kind: MomentKind, json: JSONObject?, signedIn: Boolean): MomentContent {
        if (kind == MomentKind.QUICK_KISS) return MomentContent("💋", "Thumb Kiss", "Tap to connect")
        if (!signedIn) return MomentContent("🕯️", "loversrock.", "Sign in to set up")
        if (json == null) return MomentContent("🕯️", "loversrock.", "Tap to refresh")
        if (!json.optBoolean("paired")) return MomentContent("💞", "Not paired", "Pair in the app")
        val partner = json.optString("partnerName").ifBlank { "Your partner" }

        return when (kind) {
            MomentKind.DAYS_TOGETHER -> MomentContent("❤️", "${json.optInt("daysTogether")}", "days together", "with $partner")
            MomentKind.ANNIVERSARY -> json.optJSONObject("anniversary")?.let {
                val days = it.optInt("daysRemaining")
                MomentContent("💍", if (days == 0) "Today!" else "${days}d", "until our anniversary", if (it.optInt("years") > 0) "${it.optInt("years")} years" else "")
            } ?: MomentContent("💍", "—", "Set your anniversary", "in Settings")
            MomentKind.STREAK -> MomentContent(
                "🔥", "${json.optInt("streakCount")}", "day streak",
                when {
                    !json.optBoolean("promptAnsweredToday") -> "Today's question is waiting"
                    json.optInt("streakFreezes") > 0 -> "❄️ ${json.optInt("streakFreezes")} freeze(s) banked"
                    else -> "All caught up"
                }
            )
            MomentKind.COUNTDOWN -> json.optJSONObject("nextCountdown")?.let {
                MomentContent("⏳", "${it.optInt("daysRemaining")}d", it.optString("label"), "")
            } ?: MomentContent("⏳", "—", "No countdown yet")
            MomentKind.DISTANCE -> if (json.isNull("distanceKm")) MomentContent("📍", "—", "Location sharing off")
                else json.optDouble("distanceKm").let { km ->
                    MomentContent("📍", if (km < 1) "<1 km" else "${km.toInt()} km", "apart", "from $partner")
                }
            MomentKind.PARTNER_MOOD -> json.optJSONObject("partnerMood")?.let {
                MomentContent(it.optString("emoji", "🙂"), partner, "is feeling", it.optString("text"))
            } ?: MomentContent("💭", partner, "hasn't set a mood")
            MomentKind.NEXT_DATE -> json.optJSONObject("nextDate")?.let {
                val d = it.optInt("daysRemaining")
                MomentContent("📅", it.optString("title"), if (d == 0) "today" else "in $d days", "", longValue = true)
            } ?: MomentContent("📅", "No date planned", "Swipe for ideas", "", longValue = true)
            MomentKind.LOVE_NOTE -> json.optJSONObject("latestNote")?.let {
                MomentContent("💌", it.optString("body"), "from $partner", it.optString("title"), longValue = true)
            } ?: MomentContent("💌", "No notes yet", "from $partner", "", longValue = true)
            // Only ever the fact that a secret exists — never its words.
            MomentKind.SECRET_MESSAGE -> if (json.optBoolean("secretMessageWaiting"))
                MomentContent("💌", "You have a message ❤️", "Tap to open", "", longValue = true)
                else MomentContent("🔒", "No new secrets", "Leave one for $partner", "", longValue = true)
            MomentKind.DAILY_QUESTION -> MomentContent(
                "💬", json.optString("todayQuestion").ifBlank { "No question today" }, "Today's question",
                if (json.optBoolean("promptAnsweredToday")) "You answered ✓" else "Tap to answer", longValue = true
            )
            MomentKind.QUICK_KISS -> MomentContent("💋", "Thumb Kiss", "Tap to connect")
        }
    }
}

abstract class MomentWidgetProvider : AppWidgetProvider() {
    abstract val kind: MomentKind

    override fun onUpdate(context: Context, manager: AppWidgetManager, ids: IntArray) {
        val signedIn = WidgetRepository.credentials(context) != null
        ids.forEach { render(context, manager, it, MomentRepository.cached(context), signedIn) }
        if (kind == MomentKind.QUICK_KISS) return // static, no data needed
        Thread {
            val fresh = MomentRepository.fetch(context)
            if (fresh != null) ids.forEach { render(context, manager, it, fresh, signedIn) }
        }.start()
    }

    private fun render(context: Context, manager: AppWidgetManager, widgetId: Int, json: JSONObject?, signedIn: Boolean) {
        val c = MomentFormatter.content(kind, json, signedIn)
        val views = RemoteViews(context.packageName, R.layout.widget_moment)
        views.setTextViewText(R.id.moment_icon, c.icon)
        views.setTextViewText(R.id.moment_value, c.value)
        views.setTextViewTextSize(R.id.moment_value, TypedValue.COMPLEX_UNIT_SP, if (c.longValue) 14f else 26f)
        views.setInt(R.id.moment_value, "setMaxLines", if (c.longValue) 4 else 1)
        views.setTextViewText(R.id.moment_label, c.label)
        views.setTextViewText(R.id.moment_sub, c.sub)
        views.setViewVisibility(R.id.moment_sub, if (c.sub.isBlank()) View.GONE else View.VISIBLE)

        val open = Intent(Intent.ACTION_VIEW, Uri.parse("loversrock://${kind.deepLink}")).setPackage(context.packageName)
        views.setOnClickPendingIntent(
            R.id.moment_root,
            PendingIntent.getActivity(context, kind.ordinal + 100, open, PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT)
        )
        manager.updateAppWidget(widgetId, views)
    }
}

class DaysTogetherWidget : MomentWidgetProvider() { override val kind = MomentKind.DAYS_TOGETHER }
class AnniversaryWidget : MomentWidgetProvider() { override val kind = MomentKind.ANNIVERSARY }
class StreakWidget : MomentWidgetProvider() { override val kind = MomentKind.STREAK }
class CountdownWidget : MomentWidgetProvider() { override val kind = MomentKind.COUNTDOWN }
class DistanceWidget : MomentWidgetProvider() { override val kind = MomentKind.DISTANCE }
class PartnerMoodWidget : MomentWidgetProvider() { override val kind = MomentKind.PARTNER_MOOD }
class NextDateWidget : MomentWidgetProvider() { override val kind = MomentKind.NEXT_DATE }
class LoveNoteWidget : MomentWidgetProvider() { override val kind = MomentKind.LOVE_NOTE }
class SecretMessageWidget : MomentWidgetProvider() { override val kind = MomentKind.SECRET_MESSAGE }
class DailyQuestionWidget : MomentWidgetProvider() { override val kind = MomentKind.DAILY_QUESTION }
class QuickKissWidget : MomentWidgetProvider() { override val kind = MomentKind.QUICK_KISS }

/** The shared Canvas, drawn from vector strokes. Tapping opens the Canvas. */
class CanvasWidget : AppWidgetProvider() {
    override fun onUpdate(context: Context, manager: AppWidgetManager, ids: IntArray) {
        Thread {
            val bitmap = MomentRepository.fetchCanvasBitmap(context)
            ids.forEach { id ->
                val views = RemoteViews(context.packageName, R.layout.widget_canvas)
                if (bitmap != null) {
                    views.setImageViewBitmap(R.id.canvas_image, bitmap)
                    views.setViewVisibility(R.id.canvas_image, View.VISIBLE)
                    views.setViewVisibility(R.id.canvas_empty, View.GONE)
                } else {
                    views.setViewVisibility(R.id.canvas_image, View.GONE)
                    views.setViewVisibility(R.id.canvas_empty, View.VISIBLE)
                }
                val open = Intent(Intent.ACTION_VIEW, Uri.parse("loversrock://canvas")).setPackage(context.packageName)
                views.setOnClickPendingIntent(R.id.canvas_root, PendingIntent.getActivity(context, 99, open, PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT))
                manager.updateAppWidget(id, views)
            }
        }.start()
    }
}

object MomentWidgets {
    val providers = listOf(
        DaysTogetherWidget::class.java, AnniversaryWidget::class.java, StreakWidget::class.java,
        CountdownWidget::class.java, DistanceWidget::class.java, PartnerMoodWidget::class.java,
        NextDateWidget::class.java, LoveNoteWidget::class.java, SecretMessageWidget::class.java,
        DailyQuestionWidget::class.java, QuickKissWidget::class.java, CanvasWidget::class.java,
    )

    fun refreshAll(context: Context) {
        val manager = AppWidgetManager.getInstance(context)
        for (cls in providers) {
            val ids = manager.getAppWidgetIds(ComponentName(context, cls))
            if (ids.isEmpty()) continue
            context.sendBroadcast(Intent(context, cls).apply {
                action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
                putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
            })
        }
    }
}
